README.md

```diff
@@ -1 +1,364 @@
-# NYC_Taxi_Data_Pipeline
\ No newline at end of file
+# 🚕 NYC_TAXI Data Pipeline 🚕
+
+In today's data-driven landscape, analyzing extensive datasets is essential for deriving business insights. Our Taxi Data Analytics application leverages <b>Airflow, Spark, Delta Lake, Debezium, Kafka, DBT,</b> and <b>Great Expectations</b> to convert raw taxi trip data into actionable intelligence.
+
+## 💥 Challenge: Data Integration Issues 💥
+
+At the beginning, our operations encountered considerable difficulties in integrating various data sources. The diversity in systems and formats posed challenges in consolidating and thoroughly analyzing trip data.
+
+This fragmentation led to incomplete insights and impeded our capability to make informed decisions based on data effectively. Hence, we required a robust solution to consolidate our data sources or streamline the analysis process.
+
+## 🌟 System Architecture 🌟
+
+<p align="center">
+<img src="./imgs/big-data-diagram.svg" width=100% height=100%>
```

Comment:
Some notes on the architecture:
- DBeaver is not a `query engine` but a multi-platform database tool. Trino is a distributed query engine.
- The flow from taxi to the database PostgreSQL (in STREAM) is not clear, how the data was ingested?
- In BATCH, which tools were used to ingest data into datalake? When will the batch engine (Spark) be triggered, periodically or what? Spark writes to Staging or both Staging and DWH? btw, staging can also be a dwh.
- How does GX work? Will it be triggered manually or with a schedule?

Reply:
Btw try to mention how big your data is in this project, along with processing time, to impress the employers!

Reply:
(1) Fixed! (Deleted Query Engine)
(2) For the streaming flow, I am simulating a scenario where data will be continuously downloaded into the PostgreSQL database, then CDC will be performed into Kafka and saved into the Data Lake or analyzed in real-time
(3)
- Data after being downloaded or aggregated locally using Python (or manually) will be mounted into /opt/airflow/data to perform ELT into the Data Lake (MinIO).
 - After each day, the Batch Engine will be triggered (I just added @daily to the schedule). Spark only writes data into Staging, which is just a place where data is aggregated and cleaned before data validation. The data will be transformed and loaded into separate dimension and fact tables in the Data Warehouse after the dbt step.
 - Currently, I still perform GX manually. In the future, I will add a schedule for the ETL pipeline (Spark --> Staging, Staging --> GX, Staging --> dbt --> DW)

Reply:
I think my machine won't be able to handle the dataset of ~100GB (it might, but the runtime will be relatively long). I have a screenshot of the ~100GB dataset. Can I add it to the README.md?

![data](https://github.com/trannhatnguyen2/NYC_Taxi_Data_Pipeline/assets/78005514/4bd7eabf-c5d0-43af-a2ce-56dc3b3f154a)

Reply:
100GB is not too big; consider increasing it if possible!

Reply:
I only have 5.5 GB free, I'm sorry :(